src/attributes.jl

```diff
@@ -475,6 +475,24 @@ Return the `AbstractSet` object used to define the constraint.
 """
 struct ConstraintSet <: AbstractConstraintAttribute end
 
+"""
+    ConstraintPrimalTolerance()
+
+For scalar functions `f ∈ S`, this tolerance corresponds to the amount by which
```

Comment:
This is not a proper definition. What's the distance metric? Row-wise doesn't make sense for PSD constraints, for example.

Reply:
Yes it's not a proper definition. Mainly put this PR up so we can discuss. [Gurobi](http://www.gurobi.com/documentation/7.5/refman/feasibilitytol.html) have an explicit parameter but don't specify the distance metric. [CPLEX](https://www-01.ibm.com/support/docview.wss?uid=swg21400045) have parameters for variable bound violations (again, no metric given), and some parameters to control the barrier method, but no explicit parameter for constraint tolerances.

I would expect this means something different for each (Function, Set) combination so maybe that's how it should be approached?

Reply:
Usually it would just depend on the set and not the function. We could define canonical violation metrics for each set. It would be another question to what extent solvers would support the metric.

Reply:
Something along the lines of 

```
EqualTo: [value-δ, value+δ]
LessThan: (-∞, upper+δ]
GreaterThan: [lower-δ, ∞)
Interval: [lower-δ, upper+δ]

ZeroOne: [0-δ, 0+δ] ∪ [1-δ, 1+δ]
Integers: U_{i∈Z} [i-δ, i+δ]

Reals: x∈ R^n
Zeros: {x ∈ R^n: -δ ≤ x ≤ δ}
Nonpositives: {x ∈ R^n: x ≤ δ}
Nonnegatives: {x ∈ R^n: x ≥ -δ}
Semicontinuous: [0-δ, 0+δ] ∪ [lower-δ, upper+δ]
Semiinteger: [0-δ, 0+δ] ∪ {∪_{i∈lower, lower+1, ..., upper} [i-δ, i+δ]}

SOS1: at most one variable outside the range [-δ, δ]
SOS2: at most two variables outside the range [-δ, δ], and if there are two, 
    they must be adjacent
```

Reply:
Do you know if this definition for SOS feasibility tolerance is used anywhere?

Reply:
No. It just seemed like a reasonable thing to do